python_modules/dagster/dagster_tests/core_tests/test_op_invocation.py

```diff
@@ -1333,3 +1333,212 @@ def foo(context: AssetExecutionContext):
         partition_key_range=PartitionKeyRange("2023-01-01", "2023-01-02"),
     )
     assert foo(context) == {"2023-01-01": True, "2023-01-02": True}
+
+
+def test_direct_invocation_output_metadata():
+    @asset
+    def my_asset(context):
+        context.add_output_metadata({"foo": "bar"})
+
+    @asset
+    def my_other_asset(context):
+        context.add_output_metadata({"baz": "qux"})
+
+    ctx = build_asset_context()
+
+    my_asset(ctx)
+    assert ctx.get_output_metadata("result") == {"foo": "bar"}
+
+    # context in unbound when used in another invocation. This allows the metadata to be
+    # added in my_other_asset
+    my_other_asset(ctx)
+
+
+def test_async_assets_with_shared_context():
+    @asset
+    async def async_asset_one(context):
+        assert context.asset_key.to_user_string() == "async_asset_one"
+        await asyncio.sleep(0.01)
+        return "one"
+
+    @asset
+    async def async_asset_two(context):
+        assert context.asset_key.to_user_string() == "async_asset_two"
+        await asyncio.sleep(0.01)
+        return "two"
+
+    # test that we can run two ops/assets with the same context at the same time without
+    # overriding op/asset specific attributes
+    ctx = build_asset_context()
+
+    async def main():
+        return await asyncio.gather(
+            async_asset_one(ctx),
+            async_asset_two(ctx),
+        )
+
+    with pytest.raises(
+        DagsterInvalidInvocationError,
+        match=r"This context is currently being used to execute .* The context"
+        r" cannot be used to execute another op until .* has finished executing",
+    ):
+        asyncio.run(main())
+
+
+def test_context_bound_state_non_generator():
+    @asset
+    def my_asset(context):
+        assert context._bound_properties is not None  # noqa: SLF001
+
+    ctx = build_op_context()
+    assert ctx._bound_properties is None  # noqa: SLF001
+
+    my_asset(ctx)
+    assert ctx._bound_properties is None  # noqa: SLF001
+    assert ctx._invocation_properties is not None  # noqa: SLF001
```

Comment:
maybe make some not `_` prefixed methods for these tests like  `test_is_bound` so you can doc block there why the implementation detail is under test like this